server/src/main/java/org/elasticsearch/action/admin/indices/refresh/TransportUnpromotableShardRefreshAction.java

```diff
@@ -55,6 +86,39 @@ public TransportUnpromotableShardRefreshAction(
             transportService.getThreadPool().executor(ThreadPool.Names.REFRESH)
         );
         this.indicesService = indicesService;
+        this.threadPool = threadPool;
+        this.useRefreshBlock = useRefreshBlock;
+    }
+
+    @Override
+    protected void beforeDispatchingRequestToUnpromotableShards(UnpromotableShardRefreshRequest request, ActionListener<Void> listener) {
+        if (useRefreshBlock == false) {
+            listener.onResponse(null);
+            return;
+        }
+
+        var clusterStateObserver = new ClusterStateObserver(clusterService, request.getTimeout(), logger, threadPool.getThreadContext());
+
```

Comment:
It is slightly annoying that the observer does not support it but we need to have something like:

```
if (isIndexBlockedForRefresh(request.shardId().getIndexName(), clusterStateObserver.setAndGetObservedState()) == false) {
    listener.onResponse(null);
    return;
}

Reply:
That's right and why I prefer slightly more the Runnable approach. Fixed in https://github.com/elastic/elasticsearch/pull/120642/commits/179074dacbc8b9f0942e9599333786a678533f04

Reply:
I still find this simpler than the runnable way which takes over more reponsibility from the observer. One can trick the observer by passing in initial version -1, but it is a bit ugly. We should probably add a method to it to simplify this very common case.

Reply:
Yes, I was planning on doing that. I'll open a PR shortly